File: OCRSPlayCollectionView/ViewController.m
//
//  ViewController.m
//  OCRSPlayCollectionView
//
//  Created by Raysharp666 on 2020/3/7.
//  Copyright © 2020 LyongY. All rights reserved.
//

#import "ViewController.h"
#import <OCRSPlayCollectionView-Swift.h>
#import "CollectionViewCell.h"
#import "Model.h"

@interface ViewController () <RSPlayCollectionViewDelegate>

@property (nonatomic, strong) RSArray *dataSource;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    _dataSource = [[RSArray alloc] initWith:@[
        [[Model alloc] initWithNum:1],
        [[Model alloc] initWithNum:2],
        [[Model alloc] initWithNum:3],
        [[Model alloc] initWithNum:4],
        [[Model alloc] initWithNum:5],
        [[Model alloc] initWithNum:6],
        [[Model alloc] initWithNum:7],
        [[Model alloc] initWithNum:8],
        [[Model alloc] initWithNum:9],
        [[Model alloc] initWithNum:10],
    [[Model alloc] initWithNum:11],
    [[Model alloc] initWithNum:12],
    ].mutableCopy
                   ];
    
    
    RSPlayCollectionView *collection = [[RSPlayCollectionView alloc] init:_dataSource registClass:CollectionViewCell.class spliteModel:SpliteModeFour selectedIndex:5 delegate:self emptyElement:^RSPlayModelBase * _Nonnull{
        return [[Model alloc] initWithNum:999];
    }];
    
    [self.view addSubview:collection];
    collection.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.view addConstraints:@[
        [collection.topAnchor constraintEqualToAnchor:self.view.topAnchor constant:50],
        [collection.leftAnchor constraintEqualToAnchor:self.view.leftAnchor constant:0],
        [collection.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor constant:-100],
        [collection.rightAnchor constraintEqualToAnchor:self.view.rightAnchor constant:0],
    ]];

}

- (void)rsplayCollectionView:(RSPlayCollectionView *)collectionView willDisappearArray:(NSArray<NSNumber *> *)willDisappearArray {
    NSLog(@"willDisappearArray: %@", willDisappearArray);
}

- (void)rsplayCollectionView:(RSPlayCollectionView *)collectionView didDisappearArray:(NSArray<NSNumber *> *)didDisappearArray {
    NSLog(@"didDisappearArray: %@", didDisappearArray);
}

- (void)rsplayCollectionView:(RSPlayCollectionView *)collectionView didAppearArray:(NSArray<NSNumber *> *)didAppearArray {
    NSLog(@"didAppearArray: %@", didAppearArray);
}

- (void)rsplayCollectionView:(RSPlayCollectionView *)collectionView selectIndex:(NSInteger)selectIndex {
    NSLog(@"selectIndex: %ld", (long)selectIndex);
}

- (void)rsplayCollectionView:(RSPlayCollectionView *)collectionView deselectIndex:(NSInteger)deselectIndex {
    NSLog(@"deselectIndex: %ld", (long)deselectIndex);
}

- (void)rsplayCollectionView:(RSPlayCollectionView *)collectionView selectedItem:(NSInteger)selectedIndex moveTo:(NSInteger)index {
    NSLog(@"selectedItem: %ld moveTo: %ld", (long)selectedIndex, (long)index);
}

- (void)rsplayCollectionView:(RSPlayCollectionView *)collectionView cleanIndex:(NSInteger)cleanIndex {
    Model *model = (Model *)_dataSource.array[cleanIndex];
    [model clean];
    NSLog(@"cleanIndex: %ld", (long)cleanIndex);
}

@end

File: RSPlayCollectionView/ViewController.swift
//
//  ViewController.swift
//  RSPlayCollectionView
//
//  Created by Raysharp666 on 2020/3/2.
//  Copyright © 2020 LyongY. All rights reserved.
//

import UIKit

class Model: RSPlayModelBase {
    var num: Int
    
    init(num: Int) {
        self.num = num
    }
    
    override func realExistence() -> Bool {
        num <= 10
    }
    
    override var description: String {
        "\(num)"
    }
}

class ViewController: UIViewController {
    
    var collection: RSPlayCollectionView!
    
    var array = RSArray(with: [
        Model(num: 1),
        Model(num: 2),
        Model(num: 3),
        Model(num: 4),
        Model(num: 5),
        Model(num: 6),
        Model(num: 7),
        Model(num: 8),
        Model(num: 9),
        Model(num: 10),
        Model(num: 11),
    ])
    
    override func viewDidLoad() {
        super.viewDidLoad()

        collection = RSPlayCollectionView(array, registClass: RSPlayCollectionViewCell.self, maxSpliteMode: .eight, spliteModel: .four, selectedIndex: 5, delegate: self, emptyElement: {
            Model(num: 9999)
        })
        
        collection.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(collection)
        view.addConstraints([
            collection.topAnchor.constraint(equalTo: view.topAnchor, constant: 50),
            collection.leftAnchor.constraint(equalTo: view.leftAnchor),
            collection.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -80),
            collection.rightAnchor.constraint(equalTo: view.rightAnchor),
        ])
        

    }
    
    @IBAction func buttonClick(_ sender: UIButton) {
        collection.spliteMode = RSPlayCollectionView.SpliteMode(rawValue: sender.tag) ?? .one
    }
    
    
}

extension ViewController: RSPlayCollectionViewDelegate {
    func rsplayCollectionView(_ collectionView: RSPlayCollectionView, willDisappearArray: [Int]) {
        print("will disappear: \(willDisappearArray.description)")
    }
    
    func rsplayCollectionView(_ collectionView: RSPlayCollectionView, didAppearArray: [Int]) {
        print("did appear: \(didAppearArray.description)")
    }
    
    func rsplayCollectionView(_ collectionView: RSPlayCollectionView, didDisappearArray: [Int]) {
        print("did disappear: \(didDisappearArray.description)")
    }
    
    func rsplayCollectionView(_ collectionView: RSPlayCollectionView, selectIndex: Int) {
        print("selected: \(selectIndex)")
    }
    
    func rsplayCollectionView(_ collectionView: RSPlayCollectionView, deselectIndex: Int) {
        print("deselected: \(deselectIndex)")
    }
    
    func rsplayCollectionView(_ collectionView: RSPlayCollectionView, selectedItem selectedIndex: Int, moveTo index: Int) {
        print("selectedItem: \(selectedIndex) move to: \(index)")
    }
    
    func rsplayCollectionView(_ collectionView: RSPlayCollectionView, cleanIndex: Int) {
        print("delete: \(cleanIndex)")
        let rr = array.array[cleanIndex] as! Model
        rr.num = 9999
    }
}

